File: Packages/RideSharingFeature/Sources/RideSharingFeature/DesignSystem/LiiveProgressDots.swift
import SwiftUI

public struct LiiveProgressDots: View {
    let legs: Int
    let current: Int

    public init(legs: Int, current: Int) {
        self.legs = min(max(LiiveProgressDotsLayout.firstLeg, legs), LiiveProgressDotsLayout.maxLegs)
        self.current = max(LiiveProgressDotsLayout.firstLeg, current)
    }

    public var body: some View {
        HStack(alignment: .center, spacing: LiiveProgressDotsLayout.containerSpacing) {
            ForEach(LiiveProgressDotsLayout.firstLeg...legs, id: \.self) { index in
                leg(index)
                if index < legs {
                    transfer(after: index)
                }
            }
        }
        .frame(maxWidth: .infinity)
    }

    private func leg(_ index: Int) -> some View {
        let completed = index < current
        let active = index == current

        return VStack(spacing: LiiveProgressDotsLayout.legGap) {
            Text("\(index)")
                .font(LiiveFont.caption1.weight(.bold).monospacedDigit())
                .foregroundColor(
                    completed || active ? LiiveProgressDotsLayout.activeTextColor : LiiveColor.textTertiary
                )
                .frame(width: LiiveProgressDotsLayout.legCircleSize, height: LiiveProgressDotsLayout.legCircleSize)
                .background(legColor(completed: completed, active: active))
                .clipShape(Circle())
            Text("Leg \(index)")
                .font(LiiveFont.caption2)
                .foregroundColor(LiiveColor.textSecondary)
        }
        .fixedSize()
    }

    private func transfer(after index: Int) -> some View {
        let passed = index < current

        return VStack(spacing: LiiveProgressDotsLayout.transferGap) {
            Rectangle()
                .fill(passed ? LiiveColor.success : LiiveColor.fill)
                .frame(height: LiiveProgressDotsLayout.connectorHeight)
                .clipShape(Capsule())
            Image(systemName: "arrow.triangle.swap")
                .font(.system(size: LiiveProgressDotsLayout.transferIconSize, weight: .bold))
                .foregroundColor(passed ? LiiveColor.success : LiiveColor.warning)
        }
        .frame(minWidth: LiiveProgressDotsLayout.transferMinWidth, maxWidth: .infinity)
        .padding(.bottom, LiiveProgressDotsLayout.transferBottomPadding)
    }

    private func legColor(completed: Bool, active: Bool) -> Color {
        if completed {
            return LiiveColor.success
        }
        if active {
            return LiiveColor.accent
        }
        return LiiveColor.fill
    }
}

private enum LiiveProgressDotsLayout {
    static let firstLeg = 1
    static let maxLegs = 3
    static let containerSpacing: CGFloat = 0
    static let legGap = LiiveSpacing.xs
    static let legCircleSize = LiiveSpacing.xxl
    static let transferGap = LiiveSpacing.xs - LiiveSpacing.xs2 / 2
    static let connectorHeight = LiiveSpacing.xs2
    static let transferIconSize = LiiveSpacing.m + LiiveSpacing.xs2 / 2
    static let transferMinWidth = LiiveSpacing.xxl + LiiveSpacing.xs
    static let transferBottomPadding = LiiveSpacing.l - LiiveSpacing.xs2 / 2
    static let activeTextColor = Color.white
}

File: android/app/src/main/java/com/liive/ride/designsystem/LiiveProgressDots.kt
package com.liive.ride.designsystem

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.font.FontWeight

@Composable
fun LiiveProgressDots(legs: Int, current: Int, modifier: Modifier = Modifier) {
    val boundedLegs = legs.coerceIn(LiiveProgressDotsLayout.FirstLeg, LiiveProgressDotsLayout.MaxLegs)
    val boundedCurrent = current.coerceAtLeast(LiiveProgressDotsLayout.FirstLeg)

    Row(modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
        for (index in LiiveProgressDotsLayout.FirstLeg..boundedLegs) {
            ProgressLeg(index = index, current = boundedCurrent)
            if (index < boundedLegs) {
                ProgressTransfer(passed = index < boundedCurrent, modifier = Modifier.weight(1f))
            }
        }
    }
}

@Composable
private fun ProgressLeg(index: Int, current: Int) {
    val c = LiiveTheme.colors
    val completed = index < current
    val active = index == current
    val background = when {
        completed -> c.success
        active -> c.accent
        else -> c.fill
    }

    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(LiiveProgressDotsLayout.LegGap)
    ) {
        Box(
            Modifier
                .size(LiiveProgressDotsLayout.LegCircleSize)
                .clip(CircleShape)
                .background(background),
            contentAlignment = Alignment.Center
        ) {
            Text(
                text = index.toString(),
                color = if (completed || active) LiiveProgressDotsLayout.ActiveTextColor else c.textTertiary,
                style = MaterialTheme.typography.labelMedium.tabularNumbers(),
                fontWeight = FontWeight.Bold
            )
        }
        Text("Leg $index", color = c.textSecondary, style = MaterialTheme.typography.labelSmall)
    }
}

@Composable
private fun ProgressTransfer(passed: Boolean, modifier: Modifier = Modifier) {
    val c = LiiveTheme.colors
    val color = if (passed) c.success else c.warning

    Column(
        modifier
            .defaultMinSize(minWidth = LiiveProgressDotsLayout.TransferMinWidth)
            .padding(bottom = LiiveProgressDotsLayout.TransferBottomPadding),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(LiiveProgressDotsLayout.TransferGap)
    ) {
        Box(
            Modifier
                .fillMaxWidth()
                .height(LiiveProgressDotsLayout.ConnectorHeight)
                .clip(LiiveRadius.full)
                .background(if (passed) c.success else c.fill)
        )
        Icon(
            painterResource(RideIcons.SwapHoriz),
            null,
            tint = color,
            modifier = Modifier.size(LiiveProgressDotsLayout.TransferIconSize)
        )
    }
}

private object LiiveProgressDotsLayout {
    const val FirstLeg = 1
    const val MaxLegs = 3
    val LegGap = LiiveSpacing.xs
    val LegCircleSize = LiiveSpacing.xxl
    val TransferGap = LiiveSpacing.xs - LiiveSpacing.xs2 / 2
    val ConnectorHeight = LiiveSpacing.xs2
    val TransferIconSize = LiiveSpacing.m + LiiveSpacing.xs2 / 2
    val TransferMinWidth = LiiveSpacing.xxl + LiiveSpacing.xs
    val TransferBottomPadding = LiiveSpacing.l - LiiveSpacing.xs2 / 2
    val ActiveTextColor = Color.White
}
